lib/spack/spack/test/packaging.py

```diff
@@ -0,0 +1,236 @@
+import argparse
+import spack.cmd.buildcache as buildcache
+import spack.relocate as relocate
+import platform
+import os
+import stat
+
+
+@pytest.fixture(scope='function')
+def install_mockery(tmpdir, config, builtin_mock):
+    """Hooks a fake install directory and a fake db into Spack."""
+    layout = spack.store.layout
+    db = spack.store.db
+    # Use a fake install directory to avoid conflicts bt/w
+    # installed pkgs and mock packages.
+    old_stage_path = spack.stage_path
+    spack.store.layout = YamlDirectoryLayout(str(tmpdir))
+    spack.store.db = Database(str(tmpdir))
+    spack.stage_path = join_path(tmpdir, 'stage')
+    # We use a fake package, so skip the checksum.
+    spack.do_checksum = False
+    yield
+    # Turn checksumming back on
+    spack.do_checksum = True
+    # Restore Spack's layout.
+    spack.store.layout = layout
+    spack.store.db = db
+    spack.stage_path = old_stage_path
+
+
+def fake_fetchify(url, pkg):
+    """Fake the URL for a package so it downloads from a file."""
+    fetcher = FetchStrategyComposite()
+    fetcher.append(URLFetchStrategy(url))
+    pkg.fetcher = fetcher
+
+
+@pytest.fixture(scope='function')
+def testing_gpg_directory(tmpdir):
+    old_gpg_path = spack.util.gpg.GNUPGHOME
+    spack.util.gpg.GNUPGHOME = str(tmpdir.join('gpg'))
+    yield
+    spack.util.gpg.GNUPGHOME = old_gpg_path
+
+
+def has_gnupg2():
+    try:
+        spack.util.gpg.Gpg.gpg()('--version', output=os.devnull)
+        return True
+    except Exception:
+        return False
+
+
+@pytest.mark.usefixtures('install_mockery', 'testing_gpg_directory')
+def test_packaging(mock_archive, tmpdir):
+    # tweak patchelf to only do a download
+    spec = Spec("patchelf")
+    spec.concretize()
+    pkg = spack.repo.get(spec)
+    fake_fetchify(pkg.fetcher, pkg)
+    mkdirp(join_path(pkg.prefix, "bin"))
+    patchelfscr = join_path(pkg.prefix, "bin", "patchelf")
+    f = open(patchelfscr, 'w')
+    body = """#!/bin/bash
+echo $PATH"""
+    f.write(body)
+    f.close()
+    st = os.stat(patchelfscr)
+    os.chmod(patchelfscr, st.st_mode | stat.S_IEXEC)
+
+    # Install the test package
+    spec = Spec('trivial-install-test-package')
+    spec.concretize()
+    assert spec.concrete
+    pkg = spack.repo.get(spec)
+    fake_fetchify(mock_archive.url, pkg)
+    pkg.do_install()
+
+    # Put some non-relocatable file in there
+    filename = join_path(spec.prefix, "dummy.txt")
+    with open(filename, "w") as script:
+        script.write(spec.prefix)
+
+    # Create the build cache  and
+    # put it directly into the mirror
+
+    mirror_path = join_path(tmpdir, 'test-mirror')
+    specs = [spec]
+    spack.mirror.create(
+        mirror_path, specs, no_checksum=True
+    )
+
+    parser = argparse.ArgumentParser()
+    buildcache.setup_parser(parser)
+
+    if has_gnupg2():
+        spack.util.gpg.Gpg.trust(spack.mock_gpg_keys_path + '/external.key')
+        spack.util.gpg.Gpg.create(name='test key 1', expires='0',
```

Comment:
IMO if it were called "donttrustthis.key" that would be fair warning. Currently there is only one test creating it (for binary caching anyhow) and I still feel it will take a long time. That being said I'll try this out more before making a demand for it and see how much time it takes. Are there options for generating weaker keys faster?